go-controller/pkg/ovn/controller/services/repair.go

```diff
@@ -2,154 +2,174 @@ package services
-				key := virtualIPKey(vip, svcPort.Protocol)
-				svcVIPsProtocolMap.Insert(key)
-			}
+
+	// Look for any load balancers whose Service no longer exists in the apiserver
+	staleLBs := []string{}
+	for _, lb := range existingLBs {
+		// Extract namespace + name, look to see if it exists
+		owner := lb.ExternalIDs["k8s.ovn.org/owner"]
+		namespace, name, err := cache.SplitMetaNamespaceKey(owner)
+		if err != nil || namespace == "" {
+			klog.Warningf("Service LB %#v has unreadable owner, deleting", lb)
+			staleLBs = append(staleLBs, lb.UUID)
 		}
-	}
 
-	// Reconcile with OVN state
-	// Obtain all the VIPs present in the OVN LoadBalancers
-	// and delete the ones that are not present in Kubernetes
-	for _, p := range protocols {
-		for _, lb := range ovnLBCache[p] {
-			vips, err := loadbalancer.GetLoadBalancerVIPs(lb)
-			if err != nil {
-				klog.V(4).Infof("Failed to get vips for %s load balancer %s, err: %v", p, lb, err)
-				continue
-			}
-			txn := util.NewNBTxn()
-			for vip := range vips {
-				key := virtualIPKey(vip, p)
-				// Virtual IP and protocol doesn't belong to a Kubernetes service
-				if !svcVIPsProtocolMap.Has(key) {
-					klog.Infof("Deleting non-existing Kubernetes vip %s from OVN %s load balancer %s", vip, p, lb)
-					if err := loadbalancer.DeleteLoadBalancerVIP(txn, lb, vip); err != nil {
-						klog.V(4).Infof("Failed to delete %s load balancer vips %s for %s, err: %v", p, vip, lb, err)
-					}
-				}
-			}
-			stdout, stderr, err := txn.Commit()
-			if err != nil {
-				return fmt.Errorf("error in deleting %s load balancer %s stale vips, "+
-					"stdout: %q, stderr: %q, err: %v",
-					p, lb, stdout, stderr, err)
-			}
+		_, err = r.serviceLister.Services(namespace).Get(name)
+		if apierrors.IsNotFound(err) {
+			klog.V(5).Infof("Found stale service LB %#v", lb)
+			staleLBs = append(staleLBs, lb.UUID)
 		}
 	}
 
+	// Delete those stale load balancers
+	if err := ovnlb.DeleteLBs(staleLBs); err != nil {
+		klog.Errorf("Failed to delete stale LBs: %v", err)
+		return fmt.Errorf("failed to delete stale LBs: %w", err)
+	}
+	klog.V(2).Infof("Deleted %d stale service LBs", len(staleLBs))
+
 	// Remove existing reject rules. They are not used anymore
 	// given the introduction of idling loadbalancers
-	err = acl.PurgeRejectRules(r.clusterPortGroupUUID)
+	err = acl.PurgeRejectRules(clusterPortGroupUUID)
 	if err != nil {
 		klog.Errorf("Failed to purge existing reject rules: %v", err)
 		return err
 	}
 	return nil
 }
+
+// serviceSynced is called by a ServiceController worker when it has successfully
+// applied a service.
+// If all services have successfully synced at least once, kick off
+// runAfterSync()
+func (r *repair) serviceSynced(key string) {
+	r.Lock()
+	defer r.Unlock()
+	if len(r.unsyncedServices) == 0 {
+		return
+	}
+	delete(r.unsyncedServices, key)
+	if len(r.unsyncedServices) == 0 {
+		go r.runAfterSync() // run in a goroutine so we don't block the ServiceController
```

Comment:
how the lock plays here with the goroutine?

Reply:
runAfterSync doesn't access unsyncedServices, so it's safe.

I could call it unsyncedServicesLock if that would be clearer.

Reply:
my question is if the `go r.runAfterSync()` is not affected by the lock, so the function exits and unlocks , and the goroutine escapes the lock  ... this is more a naive question/curiosity  than a comment/review

Reply:
golang have these things you know :smile:

Reply:
yeah, it does escape, scratch my comment
https://play.golang.org/p/pRXtSvwvHAU